sourced/ml/algorithms/uast_struct_to_bag.py

```diff
@@ -1,24 +1,32 @@
 import random
 from collections import defaultdict, deque
 
-from sourced.ml.algorithms.uast_ids_to_bag import FakeVocabulary
+from sourced.ml.algorithms.uast_ids_to_bag import FakeVocabulary, Uast2BagBase
 
 
-class UastStructure2BagBase:
-    def uast_to_bag(self, uast):
-        raise NotImplemented
+class Uast2StructBagBase(Uast2BagBase):
+    def __init__(self, stride, seq_len, node2index=None):
+        self._node2index = node2index if node2index is not None else FakeVocabulary()
+        self._stride = stride
+        if not isinstance(seq_len, (int, tuple, list)):
+            raise TypeError("Unexpected type of seq_len: %s" % type(seq_len))
+        self._seq_lens = [seq_len] if isinstance(seq_len, int) else seq_len
+
+    @property
+    def node2index(self):
```

Comment:
similar to `token2index` property. But here we do not have `token_parser` stage.

Reply:
but you can do the same in `__getitem__` style. 
also, it will allow you to split this algorithm from "find most similar in dictionary" one.
Also if you need any state to find the similar combination of roles (why not) it is better to have a class.

Reply:
make sense